python/ray/experimental/signal.py

```diff
@@ -0,0 +1,184 @@
+    return ray._raylet.compute_task_id(object_id)
+
+def send(signal, source_id = None):
+    """Send signal on behalf of source_id.
+    Each signal is identified by (source_id, index), where index is incremented
+    every time a signal is sent, starting from 1. Receiving this signal,
+    requires waiting on (source_id, index).
+    Args:
+        signal: signal to be sent.
+        source_id: If empty, initialize to the id of the task/actor
+                   invoking this function.
+    """
+    if source_id == None:
+        if hasattr(ray.worker.global_worker, "actor_creation_task_id"):
+            source_key = ray.worker.global_worker.actor_creation_task_id.binary()
+        else:
+            # no actors; this function must have been called from a task
+            source_key = ray.worker.global_worker.current_task_id.binary()
+    else:
+        source_key = source_id.binary()
+
+    index = ray.worker.global_worker.redis_client.incr(source_key)
+    if index < START_SIGNAL_COUNTER:
+        ray.worker.global_worker.redis_client.set(source_key, START_SIGNAL_COUNTER)
+        index = START_SIGNAL_COUNTER
+
+    object_id = _get_signal_id(ray.ObjectID(source_key), index)
+    ray.worker.global_worker.store_and_register(object_id, signal)
+
+def receive(source_ids, timeout=float('inf')):
+    """Get all signals from each source in source_ids.
+    For each source_id in source_ids, this function returns all signals
+    generated by (or on behalf of) source_id since the last receive() or
+    forget() were invoked on source_id. If this is the first call on
+    source_id, this function returns all signals generated by (or on
+    behalf of) source_id so far.
+    Args:
+        source_ids: list of source ids whose signals are returned.
+        timeout: time it receives for new signals to be generated. If none,
+                 return when timeout experies. Measured in seconds.
+    Returns:
+        The list of signals generated for each source in source_ids. They
+        are returned as a list of pairs (source_id, signal). There can be
+        more than a signal for the same source_id.
+    """
+    if not hasattr(ray.worker.global_worker, "signal_counters"):
+        ray.worker.global_worker.signal_counters = dict()
+
+    signal_counters = ray.worker.global_worker.signal_counters
+    results = []
+    previous_time = time.time()
+    remaining_time = timeout
+
+    # If we never received a signal from a source_id, initialize the
+    # signal couunter for source_id to START_SIGNAL_COUNTER.
+    for source_id in source_ids:
+        if not source_id in signal_counters:
+            signal_counters[source_id] = START_SIGNAL_COUNTER
+
+    # For each source_id compute the id of the next unread signal and store these
+    # signals in signal_ids. Also, store the reverse mapping from signals to
+    # source ids in the source_id_from_signal_id dictionary.
+    source_id_from_signal_id = dict()
+    signal_ids = []
```

Comment:
why do we need a separate list signal_ids here? This seems to be just source_id_from_signal_id.keys()